src/PetriEngine/Colored/Reduction/RedRuleDeadTransitions.cpp

```diff
@@ -6,74 +6,91 @@
         bool continueReductions = false;
         const size_t numberofplaces = red.placeCount();
-        for(uint32_t p = 0; p < numberofplaces; ++p)
-        {
-            if(red.hasTimedOut()) return false;
+        for (uint32_t p = 0; p < numberofplaces; ++p) {
+            if (red.hasTimedOut()) return false;
             Place place = red.places()[p];
-            if(place.skipped) continue;
-            if(place.inhibitor) continue;
-            if(place._pre.size() > place._post.size()) continue;
+            if (place.skipped) continue;
+            if (place.inhibitor) continue;
+            if (place._pre.size() > place._post.size()) continue;
 
             bool ok = true;
             // Check for producers without matching consumers first
-            for(uint prod : place._pre)
-            {
+            for (uint prod: place._pre) {
                 // Any producer without a matching consumer blocks this rule
-                Transition t = red.transitions()[prod];
+                const Transition &t = red.transitions()[prod];
                 auto in = red.getInArc(p, t);
-                if(in == t.input_arcs.end())
-                {
+                if (in == t.input_arcs.end()) {
                     ok = false;
                     break;
                 }
             }
 
-            if(!ok) continue;
+            if (!ok) continue;
 
             std::set<uint32_t> notenabled;
             // Out of the consumers, tally up those that are initially not enabled by place
             // Ensure all the enabled transitions that feed back into place are non-increasing on place.
-            for(uint cons : place._post)
-            {
-                Transition t = red.transitions()[cons];
-                auto in = red.getInArc(p, t);
-                if(in->expr->weight() <= place.marking.size())
-                {
-                    // This branch happening even once means notenabled.size() != consumers.size()
-                    auto out = red.getOutArc(t, p);
-                    // Only increasing loops are not ok
+            for (uint cons: place._post) {
+                const Transition &t = red.transitions()[cons];
+
+                const auto &in = red.getInArc(p, t);
+                const auto &out = red.getOutArc(t, p);
+
+                //Cheap check
+                if (in->expr->weight() <= place.marking.size()) {
                     if (out != t.output_arcs.end() && out->expr->weight() > in->expr->weight()) {
                         ok = false;
                         break;
                     }
                 }
-                else
-                {
+
+                //slightly more expensive check
+                uint32_t colorSize = getColorSize(t);
+                if (colorSize > 10000) continue;
+                if (markingEnablesInArc(place.marking, *in, t, partition, in->expr->getColors(red.colors()))) {
```

Comment:
As far as I can tell, the call to `in->expr->getColors` is unnecessary. Just passing `red.colors()` should be faster:
```suggestion
                if (markingEnablesInArc(place.marking, *in, t, partition, red.colors())) {
```

Reply:
The idea was to only consider the colors actually used in the arc, instead of looking at bindings for every color available
```
bool RedRuleDeadTransitions::markingEnablesInArc(Multiset &marking, const Arc &arc,
                                                     const Colored::Transition &transition,
                                                     PartitionBuilder &partition,
                                                     const ColorTypeMap &colors) const {
        assert(arc.input);
        NaiveBindingGenerator gen(transition, colors);
        for (const auto &binding: gen) {
```

Reply:
Yes. But I looked into what the NaiveBindingGenerator uses the type map for. It just looks up the type of tuples constituents to find the tuples type. So since it is just lookups and not iteration, then a smaller colortypemap does not make any difference.

Reply:
Ok, sounds good.

Reply:
Should i remove the getColors() function i implemented?

Reply:
Yeah, we probably don't need it, since we have `red.colors()`